DataFormats/HcalDetId/interface/HcalDetId.h

```diff
@@ -31,63 +32,261 @@ class HcalDetId : public DetId {
 
 public:
   /** Create a null cellid*/
-  HcalDetId();
+  constexpr HcalDetId() : DetId() {
+  }
   /** Create cellid from raw id (0=invalid tower id) */
-  HcalDetId(uint32_t rawid);
+  constexpr HcalDetId(uint32_t rawid) {
+    if ((DetId::Detector(rawid>>DetId::kDetOffset)&DetId::kDetMask) != Hcal) {
+      id_ = rawid;
+    } else  {
+      HcalSubdetector subdet = (HcalSubdetector)((rawid>>DetId::kSubdetOffset)&DetId::kSubdetMask);
+      if ((subdet==HcalBarrel) || (subdet==HcalEndcap) ||
+	  (subdet==HcalOuter) || (subdet==HcalForward)) {
+        id_ = newForm(rawid);
+      } else {
+        id_ = rawid;
+      }
+    }
+  }
+
   /** Constructor from subdetector, signed tower ieta,iphi,and depth */
-  HcalDetId(HcalSubdetector subdet, int tower_ieta, int tower_iphi, int depth);
+  constexpr HcalDetId(HcalSubdetector subdet, int tower_ieta, int tower_iphi, int depth) : DetId(Hcal,subdet) {
+    // (no checking at this point!)
+    id_ |= (kHcalIdFormat2) | ((depth&kHcalDepthMask2)<<kHcalDepthOffset2) |
+      ((tower_ieta>0)?(kHcalZsideMask2|(tower_ieta<<kHcalEtaOffset2)):((-tower_ieta)<<kHcalEtaOffset2)) |
+      (tower_iphi&kHcalPhiMask2);
+  }
   /** Constructor from a generic cell id */
-  HcalDetId(const DetId& id);
+  constexpr HcalDetId(const DetId& gen) {
+    if (!gen.null()) {
+      HcalSubdetector subdet=(HcalSubdetector(gen.subdetId()));
+      if (gen.det()!=Hcal ||
+	  (subdet!=HcalBarrel && subdet!=HcalEndcap &&
+	   subdet!=HcalOuter && subdet!=HcalForward &&
+	   subdet!=HcalTriggerTower && subdet!=HcalOther)) {
+        throw cms::Exception("Invalid DetId") << "Cannot initialize HcalDetId from " << std::hex << gen.rawId() << std::dec;
+      }
+      if ((subdet==HcalBarrel) || (subdet==HcalEndcap) ||
+	  (subdet==HcalOuter) || (subdet==HcalForward)) {
+        id_ = newForm(gen.rawId());
+      } else {
+        id_ = gen.rawId();
+      }
+    } else {
+      id_ = gen.rawId();
+    }
+  }
   /** Assignment from a generic cell id */
-  HcalDetId& operator=(const DetId& id);
+  constexpr HcalDetId& operator=(const DetId& gen) {
```

Comment:
Since a `DetId` is very cheap to copy (IIUC it is just a `uint32_t`) I would change this to take `DetId gen` instead of `const DetId& gen`